csrc/scheduler/normalization_utils.cpp

```diff
@@ -864,6 +871,14 @@ bool projectBufferToInputs(
   // memory to store the buffers
   int64_t max_available_buffer = getMaxRegOrSharedMemorySizeForPersistentBuffer(
       runtime_info, persistent_buffer_info.persistent_buffers);
+  NVF_ERROR(
```

Comment:
Actually, when we return `false` here, we should make sure `persistent_buffer_info.persistent_buffers` is smaller than `max_available_buffer`. Do we have that check as well?

Reply:
that is guranteed in `canScheduleRunTime`, otherwise will be rejected by the scheduler.

Reply:
Then why did the original PR fail?

Reply:
The original PR fails becase when `projected_persistent_buffer_size <= max_available_buffer < persistent_buffer_size`, we should project to inputs to lower the buffer size but the original PR used the larger buffer size due to rng op.

`canScheduleRunTime` ensures `max_available_buffer >= std::min(projected_persistent_buffer_size, persistent_buffer_size)`, when `projected_persistent_buffer_size >= persistent_buffer_size`, we can derive `max_available_buffer >= persistent_buffer_size`

Reply:
> canScheduleRunTime ensures max_available_buffer >= std::min(projected_persistent_buffer_size, persistent_buffer_size)

Why doesn't it do `max_available_buffer >= projectBufferToInputs(...) ?projected_persistent_buffer_size : persistent_buffer_size`?

Reply:
I tried to use this function to check whether we can project to inputs in run time check. But got an issue since the segmenter will simply peel off the rng op from the original fusion. Here is the original post, [link](https://github.com/NVIDIA/Fuser/pull/1885#issuecomment-1984964113).

Reply:
>  the segmenter will simply peel off the rng op from the original fusion

What do you mean? Could you give more specific info?

Reply:
The segmenter generates a fusion with a single rng op. No input to the fusion.

Error msg: ` Could not launch kernel as program could not infer i0(i0) for the buffer T5_g[ iS8{i0}, iS9{i1} ] `

The Fusion:
```
Inputs:
Outputs:
  T5_g[ iblockIdx.x576{( ceilDiv(( ceilDiv(( ceilDiv(( i0 * i1 ), 8) ), 1) ), 128) )}, iUS575{1}, iV573{8}, ithreadIdx.x577{128} ] ca_pos( 2 ) produce_pos( 2 ), bool

%kernel_math {
T4_l[ iblockIdx.x590{( ceilDiv(( ceilDiv(( ceilDiv(( i0 * i1 ), 8) ), 1) ), 128) )}, iUS589{1}, iS587{8}, ithreadIdx.x591{128} ] ca_pos( 4 )
   = rng_uniform_range({i0, i1}, double(0), double(1), float);
T296_l[ iblockIdx.x583{( ceilDiv(( ceilDiv(( ceilDiv(( i0 * i1 ), 8) ), 1) ), 128) )}, iUS582{1}, iS580{8}, ithreadIdx.x584{128} ] ca_pos( 2 ) produce_pos( 4 )
   = T4_l[ iblockIdx.x590{( ceilDiv(( ceilDiv(( ceilDiv(( i0 * i1 ), 8) ), 1) ), 128) )}, iUS589{1}, iS587{8}, ithreadIdx.x591{128} ] ca_pos( 4 )
   < double(0.80000000000000004);
T5_g[ iblockIdx.x576{( ceilDiv(( ceilDiv(( ceilDiv(( i0 * i1 ), 8) ), 1) ), 128) )}, iUS575{1}, iV573{8}, ithreadIdx.x577{128} ] ca_pos( 2 ) produce_pos( 2 )
   = Set( T296_l[ iblockIdx.x583{( ceilDiv(( ceilDiv(( ceilDiv(( i0 * i1 ), 8) ), 1) ), 128) )}, iUS582{1}, iS580{8}, ithreadIdx.x584{128} ] ca_pos( 2 ) produce_pos( 4 ), cache_op=Streaming )
}
```

do we allow fusion without inputs in the segmenter? If not we may need to fix the segmenter, if allowed we need to figure out how to use other fusion-segments to derive the size of T5.

Reply:
I did read your link, but why is this related? Using `projectBufferToInputs` should not change segmentation, right? If so, we should have seen this issue no matter if `projectBuffertoInputs` was used.

Reply:
It actually does. Here is how it works if we use `projectBufferToInputs` in `canScheduleRunTime`:
(1) It detects `rng` op and will set `project_to_inputs = false`, and when `max_available_buffer < non_project_buffer_size`, the fusion segment.
(2) The segmenter start from stratch, and merged all the exprs except the one with `rng` op, up to here, the `canScheduleRunTime` returns true because `project_to_inputs = true` and `max_available_buffer > projected_buffer_size`. However, the segmenter can't merge the `rng` op back with other exprs since it will lead to situation (1).

Reply:
I'm very confused. Are you saying we are making different projection decisions between `canScheduleRuntime` and wherever others that use `projectBufferToInputs`? Isn't it considered inconsistent?

Reply:
Previsosly it is inconssitent and that is why the original PR failed in some cases.
The current logic is:
(1) `canScheduleRuntime` assumes we can project to inputs as long as projected buffer size is smaller.
(2)  Then `projectBufferToInputs` did a further check which may disable project (as long as we still have enough space to host the buffer).

Reply:
Ideally, we should use `projectBufferToInputs` in `canScheduleRuntime`, but it caused the aforementioned segmentation issue.

Reply:
These two seem inconsistent to me:

>  It detects rng op and will set project_to_inputs = false, and when max_available_buffer < non_project_buffer_size, the fusion segment.

and 

> Then projectBufferToInputs did a further check which may disable project (as long as we still have enough space to host the buffer).

Reply:
Your first reference is the old logic in `canScheduleRuntime` and it leads to the aforementioned segmentation issue.
The second reference is the revised logic in `projectBufferToInputs`
They do inconsistgent but they do NOT exist in the same branch.

I retried the revised version of `projectBufferToInputs` and it can be directly used in `canScheduleRuntime` without leading to the segmentation issue. 
The code is updated. 
In this version, project to input or not soley comes from the function `isProjectBufferToInputs` (renamed from `projectBufferToInputs`).
Thanks for the comments!

Reply:
The segmentation issue is gone because previously I used `bool has_rng_op = ir_utils::hasOpsOfType<RNGOp>(fusion);` to check rng op. I should only check the ops from inputs to the buffer as what was done in the `projectBufferToInputs`

Reply:
> Your first reference is the old logic in canScheduleRuntime and it leads to the aforementioned segmentation issue.

What do you mean? That is just from your comment above: https://github.com/NVIDIA/Fuser/pull/1961#discussion_r1532523189. It isn't about the old logic. It should be what would happen if `projecBufferToInputs` is used.

Reply:
Sorry for the confusion.  Let me do a summary of the four different approaches. Same colored blocks indicating they are calling the same version of `projecBufferToInputs` or other functions.

![image](https://github.com/NVIDIA/Fuser/assets/116412316/959a66d7-7d8c-45b8-9fe1-451c7c30ac72)

Reply:
Thanks for the clarification. 

Since this does involve a change of logic, please run the CI tests. 

Are there any existing tests/benchmarks that result in different segmentation due to this change?

Reply:
No change in the existing benchmarks.
I am also not aware of tests, but CI has some tech issues to run, will double check after fixed.